third_party/nvidia/lib/Dialect/NVWS/IR/Ops.cpp

```
@@ -0,0 +1,190 @@
+#include "triton/Dialect/TritonGPU/IR/Attributes.h"
+#include "triton/Dialect/TritonGPU/IR/Types.h"
+#include "llvm/ADT/SmallVectorExtras.h"
+
+#define GET_OP_CLASSES
+#include "Dialect/NVWS/IR/Ops.cpp.inc"
+
+namespace mlir::triton::nvws {
+
+template <typename T>
+std::optional<Twine> verifySlice(T &origType, T &newType, size_t rank) {
+  if (!origType || !newType)
+    return "MLIR Types don't match";
+  if (origType.getElementType() != newType.getElementType() ||
+      origType.getRank() - rank != newType.getRank()) {
+    return "Ranks don't match";
+  }
+  for (size_t i = 0, e = newType.getShape().size(); i < e; i++) {
+    if (origType.getShape()[i + rank] != newType.getShape()[i])
+      return "Dimensions don't match";
+  }
+  return std::nullopt;
+}
+
+std::optional<Twine>
+ArefRegionVerify(ArefType aref,
+                 ValueTypeRange<MutableArrayRef<BlockArgument>> blockArgTypes,
+                 size_t rank) {
+  auto numBatchAxes = aref.getNumBatchAxes();
+  if (numBatchAxes ? *numBatchAxes != rank : rank == 0)
+    return "The Number of Batch axes on the aref type does not match the "
+           "number of indexes";
+  auto typeArray = aref.getBaseType();
+  if (typeArray.size() != blockArgTypes.size())
+    return "Aref has different number of arguments than region";
+  // This should probably rely on the memdescSubviewOp verifier?
+  for (auto [orig, arg] : llvm::zip(typeArray, blockArgTypes)) {
+    if (rank == 0) {
+      if (orig != arg)
+        return "MLIR Types don't match";
+    } else {
+      if (auto origT = dyn_cast<RankedTensorType>(orig)) {
+        auto argT = dyn_cast<RankedTensorType>(arg);
+        if (auto result = verifySlice(origT, argT, rank))
+          return result;
+      } else if (auto origT = dyn_cast<triton::gpu::MemDescType>(orig)) {
+        auto argT = dyn_cast<triton::gpu::MemDescType>(arg);
+        if (auto result = verifySlice(origT, argT, rank))
+          return result;
+      }
```

should this throw a verifier error if the inner type is unexpected?